### ISSUE TITLE
Every source code file should have a copyleft license notice at the top

### ISSUE THREAD

**Issue by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#312]
As per the GPL.

**Comment by stephengtuggy** [comment]
Here is a sample of what the copyleft notice should look like:

```C++
/**
 * unit_generic.cpp
 *
 * Copyright (C) 2001-2002 Daniel Horn
 * Copyright (C) 2020 pyramid3d, Stephen G. Tuggy, and other Vega Strike
 * contributors
 *
 * https://github.com/vegastrike/Vega-Strike-Engine-Source
 *
 * This file is part of Vega Strike.
 *
 * Vega Strike is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Vega Strike is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Vega Strike.  If not, see <https://www.gnu.org/licenses/>.
 */
```

Some source files already had a copyleft notice referencing GPLv2; in those cases, when I update the notice to fit this format better, I use GPLv2 as before. Otherwise, where no copyleft notice was previously included, I use GPLv3. Maybe we should just put everything in Vega-Strike-Game-Engine under GPLv3 now?

**Comment by BenjamenMeyer** [comment]
We should prioritize this

**Comment by daviewales** [comment]
This looks pretty straightforward. I'm happy to work on this. Before I do, is there an answer to @stephengtuggy's question about GPLv2 vs GPLv3?

**Comment by daviewales** [comment]
This is a list of all the file extensions in the project (the command also lists all the files without extensions, but I've omitted them below):

```zsh
❯ ls -R | sed -e '/\.\/.*[:]$/d' | sed -e 's/^.*\(\..*$\)/\1/' | sort | uniq

.0
.0b
.1
.2
.4
.5
.7z
.STANDALONE
.WAV
.a
.ac
.alp
.am
.ani
.bfxm
.bin
.blank
.bmp
.c
.check_cache
.cmake
.config
.cpp
.cpt
.csv
.cube
.cur
.d
.desktop
.dir
.dll
.dsp
.dsw
.dtd
.font
.fp
.free
.h
.hpp
.icns
.ico
.image
.in
.internal
.jpg
.json
.ld
.lex
.lib
.log
.make
.marks
.md
.mission
.mtl
.nebula
.nsi
.o
.obj
.ogg
.old
.out
.pdf
.pl
.png
.ps1
.psd
.py
.rc
.rhead
.rst
.sh
.sln
.spec
.spr
.sprite
.sxc
.system
.technique
.template
.texture
.tif
.ts
.txt
.vcproj
.vp
.wav
.xai
.xmesh
.xmesh#
.xml
.xpm
.yacc
```

**Comment by stephengtuggy** [comment]
@daviewales Excellent. Glad to hear you'll be taking this on.

I believe the existing code that had copyleft notices, was under GPLv2+, which means that we should be able to relicense them under GPLv3+ without any issues. Unless someone else answers differently, that's what I would do.

You can probably start with `.c`, `.cpp`, `.h`, `.hpp`, `.py`, `.cmake`, `.sh`, and `.ps1` for now. We can add other file types later, as needed.

**Comment by daviewales** [comment]
This will find all those:

```zsh
find -E . -regex '.*\.(c|cpp|h|hpp|py|cmake|sh|ps1)'
```

**Comment by daviewales** [comment]
This is interesting... The code in `engine/src/cmd/collide2` appears to come from the [OPCODE library](http://www.codercorner.com/Opcode.htm) by Pierre Terdiman. It doesn't have an explicit license, but it says "You’re free to use them in any commercial or non-commercial program." I'm assuming that means that the code is not under the GPL, and should be excluded from this exercise?

**Comment by BenjamenMeyer** [comment]
@stephengtuggy  I would leave it as the currently one and not move to GPL3+ at least as a starting point. We can revisit GPLv3+ in another issue/PR set.

**Comment by daviewales** [comment]
@BenjamenMeyer to be clear, any file which doesn't have an existing copyleft notice should have a **GPLv2** notice added, rather than a **GPLv3** notice?

**Comment by daviewales** [comment]
Please see PR #544 addressing this issue. 😄

**Comment by BenjamenMeyer** [comment]
@daviewales correct.

**Comment by stephengtuggy** [comment]
@BenjamenMeyer I disagree. Why go to the trouble of processing all those same files, twice? Eventually, we will want them all under GPLv3+, I expect -- and much of the project is already. Why not move to GPLv3+ all in one go?

**Comment by stephengtuggy** [comment]
Also, @daviewales please update the copyleft message to include your name, and the current year. So:

```
/*
 * <filename>
 *
 * Copyright (C) Daniel Horn
 * Copyright (C) 2020 pyramid3d and other Vega Strike contributors
 * Copyright (C) 2021 daviewales
 *
 * https://github.com/vegastrike/Vega-Strike-Engine-Source
 *
 * This file is part of Vega Strike.
 *
 * Vega Strike is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Vega Strike is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Vega Strike.  If not, see <https://www.gnu.org/licenses/>.
 */
```

**Comment by BenjamenMeyer** [comment]
> @BenjamenMeyer I disagree. Why go to the trouble of processing all those same files, twice? Eventually, we will want them all under GPLv3+, I expect -- and much of the project is already. Why not move to GPLv3+ all in one go?

1. By putting them all explicitly under the same license it's clear in the history which license should be used. This also helps prevent confusion should there ever be a challenge to the decision on which license to use.
2. Once it's all explicit then we can discuss about license changes and see what's possible, but that will also require an analysis of the code as a whole.

Most likely it won't be an issue; but we need to do the analysis first.

**Comment by stephengtuggy** [comment]
@BenjamenMeyer By definition, the plus sign `+` in `GPLv2+` means that you can relicense it under any later version of the GPL, any time. Same with `GPLv3+`. See <https://www.gnu.org/licenses/> for more information.

Further, the source is not all under the same license. Many of the files are under GPL3 already.

**Comment by BenjamenMeyer** [comment]
> @BenjamenMeyer By definition, the plus sign `+` in `GPLv2+` means that you can relicense it under any later version of the GPL, any time. Same with `GPLv3+`. See https://www.gnu.org/licenses/ for more information.

I'm quite aware of licensing. I actually quite enjoy the topic ;) (Sad I know...but it can be quite fun.)

I'd still rather keep an explicit history. As I said, we can review once it's all explicitly stated.
 
> Further, the source is not all under the same license. Many of the files are under GPL3 already.

That may be. We'll leave them alone. Even the Linux Kernel is mixed license with many contributors moving to GPL3 from GPL2. Mixed licenses aren't an issue, though requires providing both so we'll need to make sure we do that too as part of the overall review.

Either way, best to clarify first then update.

**Comment by stephengtuggy** [comment]
Reading back what I wrote earlier, it almost sounds like I was suggesting that we change the license of existing files that already have a license. To clarify, that is not what I meant.

Instead, what I was trying to say was:

1. For existing files with no license/copyleft notice: Add `GPLv3+`.
2. For existing files that have a license/copyleft notice: Leave alone for now. If and when we make substantive changes to their contents, we can feel free to relicense them under `GPLv3+`, if they are currently under `GPLv2+` explicitly.
3. For any new files that we create: Put them under `GPLv3+`.
4. Note, also, that this only applies to the Vega-Strike-Engine-Source repo. The Assets-Production repo, as a whole, is listed as being licensed under `GPLv2+`. It can stay that way for now. It's just that we may want to revisit that later.

**Comment by BenjamenMeyer** [comment]
Re: https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/312#issuecomment-915395637

@stephengtuggy thanks for clarifying

Per `1` -  it should be `GPLv2+` as even though it's not marked that's what it's effectively licensed under and should reflect that.
Per `2` - we can hold a group discussion and possibly update as a whole; but let's not do that until we have made a release with everything marked.
Per `3` - I'm fine with that.
Per `4` - we should follow up with similar stories for each of our repositories.

**Comment by daviewales** [comment]
> Also, @daviewales please update the copyleft message to include your name, and the current year. So:

@stephengtuggy I haven't written any of the code in any of these files I'm updating. I've put my name and the year in the [script I wrote](https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/544/commits/aadcc09ec98a99e6532094b30a9477d47990eb3e) to add the license text. I'm happy to be included under 'Other contributors' for the source files.

**Comment by stephengtuggy** [comment]
So. We added the copyleft notices in the master branch, but not in the 0.8.x branch. Taking care of that now.

**Comment by stephengtuggy** [comment]
Fixed for 0.8.x in #639
